lib/broadway/rate_limiter.ex

```diff
@@ -0,0 +1,68 @@
+defmodule Broadway.RateLimiter do
+  @moduledoc false
+
+  use GenServer
+  @row_name :rate_limit_counter
+
+  def start_link(opts) do
+    case Keyword.fetch!(opts, :rate_limiting) do
+      # If we don't have rate limiting options, we don't even need to start this rate
+      # limiter process.
+      nil ->
+        :ignore
+
+      rate_limiting_opts ->
+        name = Keyword.fetch!(opts, :name)
+        genserver_opts = [name: Module.concat(name, __MODULE__)]
+        GenServer.start_link(__MODULE__, {name, rate_limiting_opts}, genserver_opts)
+    end
+  end
+
+  def rate_limit(table_name, amount)
+      when is_atom(table_name) and is_integer(amount) and amount > 0 do
+    :ets.update_counter(table_name, @row_name, -amount)
+  end
+
+  def get_currently_allowed(table_name) when is_atom(table_name) do
+    :ets.lookup_element(table_name, @row_name, 2)
+  end
+
+  def table_name(broadway_name) when is_atom(broadway_name) do
+    Module.concat(broadway_name, RateLimiterETS)
+  end
+
+  @impl true
+  def init({name, rate_limiting_opts}) do
+    interval = Keyword.fetch!(rate_limiting_opts, :interval)
+    allowed = Keyword.fetch!(rate_limiting_opts, :allowed_messages)
+
+    table_name = table_name(name)
+
+    _ets = :ets.new(table_name, [:named_table, :public, :set])
+    :ets.insert(table_name, {@row_name, allowed})
+
+    _ = schedule_next_reset(interval, allowed)
+
+    {:ok, {name, table_name, interval}}
+  end
+
+  @impl true
+  def handle_info({:reset_limit, allowed}, {broadway_name, table_name, interval}) do
+    was_rate_limited? = get_currently_allowed(table_name) <= 0
+
+    true = :ets.insert(table_name, {@row_name, allowed})
+
+    if was_rate_limited? do
+      producers = Broadway.producer_names(broadway_name)
+      Enum.each(producers, &send(&1, {__MODULE__, :reset_rate_limiting}))
```

Comment:
This will crash if the producer is temporarily down. We need to do a `Process.whereis(...)` on the name every time before sending it.

Reply:
I changed the code here to read the producers right after `init/1`. Can you take a look? I don't think that the `whereis` check is needed anymore now.

Reply:
We cannot read them on `init/1` because they may restart so the name will point to a new pid and we will have them outdated here. We have to do the process.whereis dance here.